### ISSUE TITLE
Mention/notify users

### ISSUE THREAD

**Issue by Awinad** [BookStackApp/BookStack#560]
### For Feature Requests

Desired Feature:
Mention users by using "@". e.g. @username
Similar to how it is done here in github.
Allow to use it on pages and in comments.
Notify mentioned user by email.

**Comment by galaxyfeeder** [comment]
+1

**Comment by genxlee** [comment]
+1

**Comment by marcusforsberg** [comment]
+1

**Comment by yoginth** [comment]
+1

**Comment by yoginth** [comment]
yeah, this is very important to BookStack and it will be the turning point for it! please take action on this issue!

**Comment by aljawaid** [comment]
+1

**Comment by berez23** [comment]
+1

**Comment by lpupillo** [comment]
+1

**Comment by abossert** [comment]
Has there been any movement or decision on this?  I would very much like to see this feature.

**Comment by ssddanbrown** [comment]
@abossert Not as of yet.

If people could use the thumbs-up emoji on the original post instead of the `+1` comments it would give the maintainers a much better idea of support for an issue like this.

**Comment by akurzawa** [comment]
+1 :) whenI craete page I would like to inform somehow for whom the page was created.

**Comment by omran987** [comment]
+1

**Comment by dinu1s** [comment]
(Y)

**Comment by sahilbhagat09** [comment]
Hi, has there been any update on this feature @ssddanbrown?

**Comment by YuntiDevOps** [comment]
+1

**Comment by brightcolor** [comment]
+2

**Comment by acrafts0** [comment]
Just :+1: 'd it as well. This may be a blocker for my group, otherwise really enjoy the workflows so far. Pitting this versus Confluence and this is a major feature difference (IMO).

**Comment by FalconWorks** [comment]
Also interested if this will get featured or not

**Comment by mschoon85** [comment]
@ssddanbrown Any idea if this feature will be build into bookstack? The owner and/or the creator of a page would like to receive e-mail notifications whenever a comment is posted. I think I would be referring to #1812

**Comment by BloodyIron** [comment]
This feature has been requested for implementation in my company, so adding that to the discussion. Can we have this added please? :)

**Comment by akahan989** [comment]
+1

**Comment by nothing-fr** [comment]
For the moment I use the webhook system to notify users of a new comment, I use n8n for the processing workflow.

It's ok for a basic use, but I'm missing some information in the json about the comments, like if it's a reply to another user a key/value "replied_to: user_id" would be welcome.
Because in this case, it's always the creator of the page who will receive the notification and not the person to whom the answer is intended.

It would be very welcome to have an internal notification system in bookstack.

**Comment by moeman2323** [comment]
Is there any movement on this request?

**Comment by lommes** [comment]
This seems to be still on the roadmap. Dan mentioned this in the [December video](https://youtu.be/ggnTbKmHkRA?si=W7K_IR8ku3PW2koI&t=487) and said that the WYSIWYG editor for the comment system (which is now available) is a prerequisite to implement it

**Comment by Grant-Smith-74** [comment]
+1 for user mentions

**Comment by quanganh206** [comment]
@lommes I do not think it native active in core app. Do we need to try some hack here?

**Comment by FabricePh** [comment]
Hello, I propose a code from my custom theme to mention @username

I guess, with Webhooks, it's possible to listen `new|edit page` to read @username and send a email or store in new table to have a notification page for current user/admin

custom_theme/functions.php
```php
<?php

use BookStack\Entities\Controllers as EntityControllers;
use BookStack\Facades\Theme;
use BookStack\Theming\ThemeEvents;
use Illuminate\Support\Facades\Route;

use Custom\Controllers as CustomControllers;

Theme::listen(ThemeEvents::ROUTES_REGISTER_WEB, function ($app) {
    Route::middleware('auth')->group(function () {
        Route::get('/api/users/search', [CustomControllers\UserController::class, 'searchUsers']);
    });
});
```

custom_theme/Controllers/UserController.php
```php
<?php

namespace Custom\Controllers;

use BookStack\Http\Controller;
use BookStack\Users\Models\User;
use Illuminate\Http\Request;

class UserController extends Controller
{
    public function searchUsers(Request $request)
    {
        $query = $request->get('query', '');
        $users = User::where('name', 'like', '%' . $query . '%')->limit(10)->get(['id', 'name']);
        return response()->json($users);
    }
}
```

custom-script.js
```javascript
function getCaretCoordinatesInTinyMCE(editor) {
  const selection = editor.selection.getRng(); 
  if (!selection) {
    return null;
  }

  const rangeRect = selection.getBoundingClientRect();

  const iframe = editor.iframeElement;
  const iframeRect = iframe.getBoundingClientRect();

  return {
    top: iframeRect.top + rangeRect.bottom + window.scrollY,
    left: iframeRect.left + rangeRect.left + window.scrollX,
  };
}

document.addEventListener("editor-tinymce::setup", (event) => {
  const editor = event.detail.editor; 

  const autocompleteBox = document.createElement("div");
  autocompleteBox.style.position = "absolute";
  autocompleteBox.style.zIndex = "1000";
  autocompleteBox.style.display = "none";
  autocompleteBox.classList.add("autocomplete-box");
  document.body.appendChild(autocompleteBox);

  editor.on("keyup", async function (e) {
    const content = editor.getContent();
    const caretCoordinates = getCaretCoordinatesInTinyMCE(editor);
    autocompleteBox.style.left = `${caretCoordinates.left}px`;
    autocompleteBox.style.top = `${caretCoordinates.top}px`;

    const selection = editor.selection;
    const range = selection.getRng();
    const textBeforeCaret = range.startContainer.textContent.slice(0, range.startOffset);

    const matchMention = textBeforeCaret.match(/@([a-zA-Z0-9\._\s]+)/);

    if (matchMention) {
      const query = matchMention[1];
      if (query.length > 0) {
        const response = await fetch(`/api/users/search?query=${query}`);
        const users = await response.json();

        autocompleteBox.style.display = "none";

        if (users.length > 0) {
          autocompleteBox.style.display = "block";
          autocompleteBox.innerHTML = users.map((user) => `<div class="mention" data-username="@${user.name}" data-userid="${user.id}">${user.name}</div>`).join("");

          autocompleteBox.querySelectorAll(".mention").forEach((item) => {
            item.addEventListener("click", function () {
              const userid = this.getAttribute("data-userid");
              const username = this.getAttribute("data-username");

              const mentionSpan = range.startContainer.parentNode.closest("span.mention");
              if (mentionSpan) {
                mentionSpan.setAttribute("data-userid", userid);
                mentionSpan.innerHTML = username;
              } else {
                const mentionHTML = `<span class="mention" data-userid="${userid}">${username}</span>`;
                editor.setContent(content.replace(matchMention[0], mentionHTML + " "));
              }

              autocompleteBox.style.display = "none";
            });
          });
        }
      } else {
        autocompleteBox.style.display = "none";
      }
    } else {
      autocompleteBox.style.display = "none";
    }
  });

  document.addEventListener("click", function () {
    autocompleteBox.style.display = "none";
  });
});
```

custom-style.css
```css
.mention {
  color: #007bff;
  font-weight: bold;
  cursor: pointer;
}

.mention:hover {
  text-decoration: underline;
}

.autocomplete-box {
  background: white;
  padding: 2px;
  border: 1px solid #ccc;
}
```

**Comment by davidejalexander** [comment]
+1 for this feature @username whenever it is possible to get it onto the work list.

**Comment by akurzawa** [comment]
I came here to have a LOL, that's the basic functionality, and in about two years, it will be 10 years since users have been asking for this feature, and it's still not implemented.

**Comment by WiredWonder** [comment]
+1 we want this too.

**Comment by michealrm** [comment]
+1, would make great use out of e-mail notifications and is necessary for efficient collaboration

**Comment by RajaRoshaan** [comment]
Is this feature implemented or not? Along with this, if someone adds a comment to someone else's page then do they get an email notification or something? Without it, we would never know if someone added a comment without check the mage manually.

**Comment by ssddanbrown** [comment]
> Is this feature implemented or not?

@RajaRoshaan No, hence why this issue is still open.
Having a WYSIWYG editor for comments was a requirement for this, which did get added, but we then had to move editors, which is a pending requirement (should be something I do sometime in the next few feature releases).

> Along with this, if someone adds a comment to someone else's page then do they get an email notification or something?

There are notification preference options for comments on owned pages, as well as comment replies.
These need to be enabled in the user's "My Account" area, and the user will need the "Receive/manage notifications" system permission in one of their roles for these options to show.

**Comment by prohtex** [comment]
> Hello, I propose a code from my custom theme to mention @username
> 
> I guess, with Webhooks, it's possible to listen `new|edit page` to read @username and send a email or store in new table to have a notification page for current user/admin
> 
> custom_theme/functions.php
> 
> <?php
> 
> use BookStack\Entities\Controllers as EntityControllers;
> use BookStack\Facades\Theme;
> use BookStack\Theming\ThemeEvents;
> use Illuminate\Support\Facades\Route;
> 
> use Custom\Controllers as CustomControllers;
> 
> Theme::listen(ThemeEvents::ROUTES_REGISTER_WEB, function ($app) {
>     Route::middleware('auth')->group(function () {
>         Route::get('/api/users/search', [CustomControllers\UserController::class, 'searchUsers']);
>     });
> });
> custom_theme/Controllers/UserController.php
> 
> <?php
> 
> namespace Custom\Controllers;
> 
> use BookStack\Http\Controller;
> use BookStack\Users\Models\User;
> use Illuminate\Http\Request;
> 
> class UserController extends Controller
> {
>     public function searchUsers(Request $request)
>     {
>         $query = $request->get('query', '');
>         $users = User::where('name', 'like', '%' . $query . '%')->limit(10)->get(['id', 'name']);
>         return response()->json($users);
>     }
> }
> custom-script.js
> 
> function getCaretCoordinatesInTinyMCE(editor) {
>   const selection = editor.selection.getRng(); 
>   if (!selection) {
>     return null;
>   }
> 
>   const rangeRect = selection.getBoundingClientRect();
> 
>   const iframe = editor.iframeElement;
>   const iframeRect = iframe.getBoundingClientRect();
> 
>   return {
>     top: iframeRect.top + rangeRect.bottom + window.scrollY,
>     left: iframeRect.left + rangeRect.left + window.scrollX,
>   };
> }
> 
> document.addEventListener("editor-tinymce::setup", (event) => {
>   const editor = event.detail.editor; 
> 
>   const autocompleteBox = document.createElement("div");
>   autocompleteBox.style.position = "absolute";
>   autocompleteBox.style.zIndex = "1000";
>   autocompleteBox.style.display = "none";
>   autocompleteBox.classList.add("autocomplete-box");
>   document.body.appendChild(autocompleteBox);
> 
>   editor.on("keyup", async function (e) {
>     const content = editor.getContent();
>     const caretCoordinates = getCaretCoordinatesInTinyMCE(editor);
>     autocompleteBox.style.left = `${caretCoordinates.left}px`;
>     autocompleteBox.style.top = `${caretCoordinates.top}px`;
> 
>     const selection = editor.selection;
>     const range = selection.getRng();
>     const textBeforeCaret = range.startContainer.textContent.slice(0, range.startOffset);
> 
>     const matchMention = textBeforeCaret.match(/@([a-zA-Z0-9\._\s]+)/);
> 
>     if (matchMention) {
>       const query = matchMention[1];
>       if (query.length > 0) {
>         const response = await fetch(`/api/users/search?query=${query}`);
>         const users = await response.json();
> 
>         autocompleteBox.style.display = "none";
> 
>         if (users.length > 0) {
>           autocompleteBox.style.display = "block";
>           autocompleteBox.innerHTML = users.map((user) => `<div class="mention" data-username="@${user.name}" data-userid="${user.id}">${user.name}</div>`).join("");
> 
>           autocompleteBox.querySelectorAll(".mention").forEach((item) => {
>             item.addEventListener("click", function () {
>               const userid = this.getAttribute("data-userid");
>               const username = this.getAttribute("data-username");
> 
>               const mentionSpan = range.startContainer.parentNode.closest("span.mention");
>               if (mentionSpan) {
>                 mentionSpan.setAttribute("data-userid", userid);
>                 mentionSpan.innerHTML = username;
>               } else {
>                 const mentionHTML = `<span class="mention" data-userid="${userid}">${username}</span>`;
>                 editor.setContent(content.replace(matchMention[0], mentionHTML + " "));
>               }
> 
>               autocompleteBox.style.display = "none";
>             });
>           });
>         }
>       } else {
>         autocompleteBox.style.display = "none";
>       }
>     } else {
>       autocompleteBox.style.display = "none";
>     }
>   });
> 
>   document.addEventListener("click", function () {
>     autocompleteBox.style.display = "none";
>   });
> });
> custom-style.css
> 
> .mention {
>   color: #007bff;
>   font-weight: bold;
>   cursor: pointer;
> }
> 
> .mention:hover {
>   text-decoration: underline;
> }
> 
> .autocomplete-box {
>   background: white;
>   padding: 2px;
>   border: 1px solid #ccc;
> }

This is working pretty well for me. @ssddanbrown love the notifications on owned pages, but in many cases we are @ mentioning users in pages themselves, and it would be amazing for them to receive an email

**Comment by Kofl** [comment]
+1

**Comment by cittadhammo** [comment]
+1

**Comment by david-prv** [comment]
+1

**Comment by ssddanbrown** [comment]
With the merge of #5944, mentions in comments has now been added and will be part of the next feature release.
I'm going to therefore close this off.

I'm aware this issue was a bit wider though, with the ability to mention from pages/content, bit it's hard to distinguish that need relative to mentions in comments, so please feel free to open/support new specific issues for mentions elsewhere.

Also, from reading some of the above, some of the support here has been for general new-comment notifications (rather than mentions) which have been available since v23.08 as detailed here: https://www.bookstackapp.com/blog/bookstack-release-v23-08/#content-notification-system
